### PR TITLE
Use `CMakeDeps` and `CMakeToolchain` as Generators on Conan

### PR DESCRIPTION
**Pull request by Nerixyz**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description

The [`cmake`](https://docs.conan.io/en/latest/reference/generators/cmake.html) generator is marked as deprecated - [`CMakeDeps`](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmakedeps.html) and [`CMakeToolchain`](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html) are recommended. This PR migrates to these generators.

### Migration Notes

0. Make sure you're using at least version 1.33 (Jan-2021).
1. Rerun `conan install`
2. Remove `-DUSE_CONAN` from your CMake arguments, and add `-DCMAKE_TOOLCHAIN_FILE="conan_toolchain.cmake"`.
3. CMake might give you an error similar to `Generator <generator> does not support platform specification...`. If that's the case, set `conf.tools.cmake.cmaketoolchain:generator` in `~/.conan/profiles/<your-profile>` to your generator. You can do this from the command line: `conan profile update conf.tools.cmake.cmaketoolchain:generator=<generator> <profile>`.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -73,8 +64,9 @@ endif ()
 find_package(Sanitizers)
 
 # Find boost on the system
-find_package(Boost REQUIRED)
-find_package(Boost COMPONENTS random)
+# `OPTIONAL_COMPONENTS random` is required for vcpkg builds to link.
+# `OPTIONAL` is required, because conan doesn't set `boost_random_FOUND`.
```
Comment (Nerixyz) [inline CMakeLists.txt:68]: I'll mention https://github.com/conan-io/conan-center-index/issues/11085 here.

**Review by pajlada — APPROVED** [review]
Looks good to me, let me know when you're ready to have this merged in :+1:

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -76,29 +76,34 @@ jobs:
...
         run: |
           choco install conan -y
 
-      - name: Enable Developer Command Prompt
+      - name: Setup Conan (Windows)
         if: startsWith(matrix.os, 'windows')
-        uses: ilammy/msvc-dev-cmd@v1.12.0
+        run: |
+          conan profile new --detect --force default
+          conan profile update conf.tools.cmake.cmaketoolchain:generator=Ninja default
```
Comment (pajlada) [inline .github/workflows/build.yml:92]: Change generator back to NMake

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -76,29 +76,34 @@ jobs:
         if: startsWith(matrix.os, 'windows')
         run: echo "C:\Program Files\Conan\conan\" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
 
+      - name: Enable Developer Command Prompt
+        if: startsWith(matrix.os, 'windows')
+        uses: ilammy/msvc-dev-cmd@v1.12.0
```
Comment (pajlada) [inline .github/workflows/build.yml:81]: Move this back to under `Install dependencies` unless you figure out this is actually necessary

**Review by pajlada — CHANGES_REQUESTED** [review]

**Review by pajlada — APPROVED** [review]

**Pull request merged**